File: auth-backend/core/services/auth/permission_service.py
"""
Permission Service
Fine-grained permission checking and management
"""
import logging
from typing import List, Optional
from core.domain.auth.permission import Permission, PermissionAction
from core.domain.auth.app_user import AppUser
# REMOVED: from core.domain.auth.user_role import UserRole (roles now in WorkspaceMember)
from core.interfaces.secondary.permission_repository_interface import IPermissionRepository

logger = logging.getLogger(__name__)


class PermissionService:
    """
    Service for fine-grained permission management.
    
    **Design Note**: resource_type is a free string (not enum) to allow
    maximum flexibility. Each system can define its own resource types.
    """
    
    def __init__(self, repository: IPermissionRepository):
        self.repository = repository
    
    async def has_permission(
        self,
        user: AppUser,
        resource_type: str,  # String livre! ✨
        action: PermissionAction,
        resource_id: Optional[str] = None
    ) -> bool:
        """
        Check if user has permission to perform action on resource.
        
        Args:
            user: User to check
            resource_type: Type of resource
            action: Action to perform
            resource_id: Specific resource ID (optional)
            
        Returns:
            True if user has permission
        """
        # REMOVED: Global admin bypass (admins are now per-workspace)
        # TODO: Implement workspace-aware admin bypass if needed
        
        # Get user's permissions for this resource type
        # Note: client_id needs to be passed explicitly (no longer in AppUser)
        client_id = getattr(user, 'client_id', 'default-client')
        permissions = await self.repository.find_by_user_and_resource_type(
            user.id, client_id, resource_type
        )
        
        # Check if any permission allows this action
        for permission in permissions:
            if permission.allows(action, resource_id):
                logger.info(f"Permission granted for user {user.id}: {action} on {resource_type}/{resource_id}")
                return True
        
        logger.warning(f"Permission denied for user {user.id}: {action} on {resource_type}/{resource_id}")
        return False
    
    async def grant_permission(
        self,
        granter: AppUser,
        user_id: str,
        resource_type: str,  # String livre! ✨
        action: PermissionAction,
        resource_id: Optional[str] = None
    ) -> Permission:
        """
        Grant permission to user.
        
        Args:
            granter: User granting permission (must be admin or have MANAGE permission)
            user_id: User receiving permission
            resource_type: Type of resource
            action: Action to allow
            resource_id: Specific resource ID (None = all resources)
            
        Returns:
            Created permission
            
        Raises:
            ValueError: If granter doesn't have permission to grant
        """
        # REMOVED: Global admin check (admins are now per-workspace)
        # Check if granter has MANAGE permission
        has_manage = await self.has_permission(
            granter, resource_type, PermissionAction.MANAGE, resource_id
        )
        if not has_manage:
            raise ValueError("You don't have permission to grant permissions")
        
        # Create permission
        # Note: client_id needs to be passed explicitly (no longer in AppUser)
        client_id = getattr(granter, 'client_id', 'default-client')
        
        permission = Permission(
            user_id=user_id,
            client_id=client_id,
            resource_type=resource_type,
            resource_id=resource_id,
            action=action,
            granted_by=granter.id
        )
        
        saved = await self.repository.save(permission)
        logger.info(f"Permission granted by {granter.id} to {user_id}: {action} on {resource_type}/{resource_id}")
        
        return saved
    
    async def revoke_permission(
        self,
        revoker: AppUser,
        permission_id: str
    ) -> bool:
        """Revoke permission (admin or granter only)"""
        permission = await self.repository.find_by_id(permission_id)
        
        if not permission:
            return False
        
        # REMOVED: Global admin check (admins are now per-workspace)
        # Only the granter or a workspace admin can revoke
        if permission.granted_by != revoker.id:
            # TODO: Check if revoker is workspace admin
            raise ValueError("You don't have permission to revoke this permission")
        
        result = await self.repository.delete(permission_id)
        logger.info(f"Permission {permission_id} revoked by {revoker.id}")
        
        return result
    
    async def list_user_permissions(
        self,
        user_id: str,
        client_id: str
    ) -> List[Permission]:
        """List all permissions for user"""
        return await self.repository.find_by_user(user_id, client_id)

File: auth-frontend/src/core/services/permission/permissionService.ts
/**
 * Permission Service Implementation
 * Business logic for permission management
 */

import type { IPermissionService } from '../../interfaces/primary/IPermissionService';
import type { IPermissionRepository } from '../../interfaces/secondary/IPermissionRepository';
import type {
  Permission,
  PermissionAction,
  GrantPermissionRequest,
} from '../../domain/permission';
import { PermissionChecker } from '../../domain/permission';
import { BusinessValidationError } from '../../domain/errors';

export class PermissionService implements IPermissionService {
  private readonly repository: IPermissionRepository;

  constructor(repository: IPermissionRepository) {
    this.repository = repository;
  }

  /**
   * Grant permission to user
   */
  async grantPermission(request: GrantPermissionRequest): Promise<Permission> {
    // Validation
    if (!request.user_id) {
      throw new BusinessValidationError('User ID is required');
    }

    if (!request.resource_type || request.resource_type.trim().length === 0) {
      throw new BusinessValidationError('Resource type is required');
    }

    if (!request.action) {
      throw new BusinessValidationError('Action is required');
    }

    // Normalize resource type
    request.resource_type = request.resource_type.toLowerCase().trim();

    return this.repository.grantPermission(request);
  }

  /**
   * List user's permissions
   */
  async listUserPermissions(userId: string): Promise<Permission[]> {
    if (!userId) {
      throw new BusinessValidationError('User ID is required');
    }

    return this.repository.listUserPermissions(userId);
  }

  /**
   * Revoke permission
   */
  async revokePermission(permissionId: string): Promise<void> {
    if (!permissionId) {
      throw new BusinessValidationError('Permission ID is required');
    }

    return this.repository.revokePermission(permissionId);
  }

  /**
   * Check if user has specific permission
   */
  async hasPermission(
    userId: string,
    resourceType: string,
    action: PermissionAction,
    resourceId?: string
  ): Promise<boolean> {
    if (!userId || !resourceType || !action) {
      return false;
    }

    try {
      const permissions = await this.repository.listUserPermissions(userId);
      return PermissionChecker.hasPermission(
        permissions,
        resourceType.toLowerCase(),
        action,
        resourceId
      );
    } catch (error) {
      console.error('Error checking permission:', error);
      return false;
    }
  }

  /**
   * Check if user can manage resource (has MANAGE permission)
   */
  async canManageResource(
    userId: string,
    resourceType: string,
    resourceId?: string
  ): Promise<boolean> {
    return this.hasPermission(userId, resourceType, 'manage', resourceId);
  }
}

